haskell/src/Stablecoin/Client/Parser.hs

```diff
@@ -0,0 +1,683 @@
+-- SPDX-FileCopyrightText: 2020 tqtezos
+-- SPDX-License-Identifier: MIT
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE ApplicativeDo #-}
+
+module Stablecoin.Client.Parser
+  ( ClientArgs(..)
+  , GlobalOptions(..)
+  , ClientArgsRaw(..)
+  , DeployContractOptions(..)
+  , TransferOptions(..)
+  , GetBalanceOfOptions(..)
+  , UpdateOperatorsOptions(..)
+  , IsOperatorOptions(..)
+  , ConfigureMinterOptions(..)
+  , RemoveMinterOptions(..)
+  , MintOptions(..)
+  , BurnOptions(..)
+  , TransferOwnershipOptions(..)
+  , ChangeMasterMinterOptions(..)
+  , ChangePauserOptions(..)
+  , SetTransferlistOptions(..)
+  , GetMintingAllowanceOptions(..)
+  , stablecoinClientInfo
+  ) where
+
+import qualified Data.List.NonEmpty as NonEmpty
+import Data.Version (showVersion)
+import Michelson.Text (MText)
+import Morley.CLI (mTextOption)
+import Morley.Client (AddressOrAlias(..), MorleyClientConfig, clientConfigParser)
+import qualified Options.Applicative as Opt
+import Options.Applicative.Help.Pretty (Doc, linebreak)
+import Paths_stablecoin (version)
+import Tezos.Address (parseAddress)
+import Util.CLI (HasCLReader(getMetavar, getReader), mkCLOptionParser)
+import Util.Named ((:!), (.!))
+
+import Stablecoin.Client.Impl (UpdateOperatorData(..))
+
+data ClientArgs = ClientArgs
+  { caMorleyClientConfig :: MorleyClientConfig
+  , caGlobalOptions :: GlobalOptions
+  , caClientArgsRaw :: ClientArgsRaw
+  }
+  deriving stock Show
+
+data GlobalOptions = GlobalOptions
+  { goUser :: AddressOrAlias
+  , goContract :: AddressOrAlias
+  }
+  deriving stock Show
+
+data ClientArgsRaw
+  = CmdDeployContract DeployContractOptions
+  -- FA2 entrypoints
+  | CmdTransfer TransferOptions
+  | CmdGetBalanceOf GetBalanceOfOptions
+  | CmdUpdateOperators UpdateOperatorsOptions
+  | CmdIsOperator IsOperatorOptions
+  | CmdPermissionsDescriptor
+  -- Stablecoin entrypoints
+  | CmdPause
+  | CmdUnpause
+  | CmdConfigureMinter ConfigureMinterOptions
+  | CmdRemoveMinter RemoveMinterOptions
+  | CmdMint MintOptions
+  | CmdBurn BurnOptions
+  | CmdTransferOwnership TransferOwnershipOptions
+  | CmdAcceptOwnership
+  | CmdChangeMasterMinter ChangeMasterMinterOptions
+  | CmdChangePauser ChangePauserOptions
+  | CmdSetTransferlist SetTransferlistOptions
+  -- Query Commands
+  | CmdGetBalance
+  | CmdGetPaused
+  | CmdGetContractOwner
+  | CmdGetPendingContractOwner
+  | CmdGetMasterMinter
+  | CmdGetPauser
+  | CmdGetTransferlist
+  | CmdGetTotalSupply
+  | CmdGetMintingAllowance GetMintingAllowanceOptions
+  | CmdGetTokenMetadata
+  deriving stock Show
+
+data DeployContractOptions = DeployContractOptions
+  { dcoMasterMinter :: AddressOrAlias
+  , dcoContractOwner :: AddressOrAlias
+  , dcoPauser :: AddressOrAlias
+  , dcoTransferlist :: AddressOrAlias
```

Comment:
Why is transferlist not optional? It's optional in the contract's storage.

Reply:
I was under the impression it was mandatory when I first began working on this issue, but now I don't think it needs to be anymore. I'll make this change.

Reply:
Done